pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -1625,6 +1625,30 @@ class Timestamp(_Timestamp):
             # When year, month or day is not given, we call the datetime
             # constructor to make sure we get the same error message
             # since Timestamp inherits datetime
+
+            if year is not None and not (0 < year < 10_000):
+                # GH#52091
+                obj = cls(
+                    year=1970,
+                    month=month,
+                    day=day,
+                    hour=hour,
+                    minute=minute,
+                    second=second,
+                    microsecond=microsecond,
+                    nanosecond=nanosecond,
+                    fold=fold,
+                    tz=tz,
+                    tzinfo=tzinfo,
+                    unit=unit,
+                )
+                if nanosecond is not None and nanosecond != 0:
+                    raise OutOfBoundsDatetime(
+                        f"Cannot construct a Timestamp with year={year} and "
+                        "non-zero nanoseconds"
+                    )
+                return obj.as_unit("us").replace(year=year)
```

Comment:
What if the `year` goes beyond the `us` max year but within the `ms` max year?

Reply:
Good point.  And same for seconds.

Now you've got me thinking about value-dependent behavior.  We should probably get "us" unit whenever nanos is zero-or-None regardless of year.  Same for getting "ms" if micros is (zero % 1000)-or-None, etc